cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/PullOperation.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.sourcecontrol;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.operation.LongRunningOperation;
+import io.cdap.cdap.internal.app.sourcecontrol.PullOperationRequest.AppToPull;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.operationrun.OperationError;
+import io.cdap.cdap.proto.operationrun.OperationMeta;
+import io.cdap.cdap.proto.operationrun.OperationResource;
+import io.cdap.cdap.proto.operationrun.OperationResourceScopedError;
+import io.cdap.cdap.sourcecontrol.operationrunner.InMemorySourceControlOperationRunner;
+import io.cdap.cdap.sourcecontrol.operationrunner.MultiPullAppOperationRequest;
+import java.lang.reflect.Type;
+import java.time.Clock;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Defines operation for doing SCM Pull for connected repositories. Supports doing pull for multiple
+ * apps. The pulled spec will be deployed through
+ * {@link io.cdap.cdap.internal.app.services.ApplicationLifecycleService} while running in
+ * app-fabric and through deploy apis when running in task workers.
+ */
+public class PullOperation implements LongRunningOperation<PullOperationRequest> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PullOperation.class);
+
+  private final InMemorySourceControlOperationRunner scmOpRunner;
+  private final SourceControlHelper helper;
+
+  private final Set<ApplicationId> deployed;
+
+  @Inject
+  PullOperation(InMemorySourceControlOperationRunner scmOpRunner, SourceControlHelper helper) {
```

Comment:
shouldn't this be a SourceControlOperationRunner instead of the InMemory one?

Reply:
As mentioned in the description for easy reuse of existing code we are putting the new multi push and pull logic in `InMemorySourceControlOperationRunner`, rather than refactoring out into a new class, which will be done in future.

The `PullOperation` would always use `InMemorySourceControlOperationRunner` as the remote runner just creates a task which we do not want.